pageserver/src/walredo.rs

```diff
@@ -587,419 +692,889 @@ impl<C: CommandExt> CloseFileDescriptors for C {
+                .map_err(StdoutTaskError::ReadFailed)?;
+            if read == 0 {
+                return Err(StdoutTaskError::StdoutClosed);
+            }
+            if buffer.len() < 8192 {
+                continue;
+            }
+            return Ok(());
+        }
     }
-}
 
-impl NoLeakChild {
-    fn spawn(command: &mut Command) -> io::Result<Self> {
-        let child = command.spawn()?;
-        Ok(NoLeakChild { child: Some(child) })
-    }
+    // TODO: do these pages are put it in a cache? if not, could use a larger buffer
+    let mut result_rx = result_rx;
 
-    fn kill_and_wait(mut self) {
-        let child = match self.child.take() {
-            Some(child) => child,
-            None => return,
+    // the +1 is for normally reading and clearing out the readyness for the stdout. tokio does not
+    // expose an api for this, so we need to have this "hack". this buffer is reused between reads.
+    let mut page_buf = Vec::with_capacity(8192 + 1);
+
+    loop {
+        let (completion, timeout_at) = match result_rx.recv().await {
+            Some(t) => t,
+            None => return Err("stdout: all requests processed, ready for shutdown"),
         };
-        Self::kill_and_wait_impl(child);
-    }
-
-    #[instrument(skip_all, fields(pid=child.id()))]
-    fn kill_and_wait_impl(mut child: Child) {
-        let res = child.kill();
-        if let Err(e) = res {
-            // This branch is very unlikely because:
-            // - We (= pageserver) spawned this process successfully, so, we're allowed to kill it.
-            // - This is the only place that calls .kill()
-            // - We consume `self`, so, .kill() can't be called twice.
-            // - If the process exited by itself or was killed by someone else,
-            //   .kill() will still succeed because we haven't wait()'ed yet.
-            //
-            // So, if we arrive here, we have really no idea what happened,
-            // whether the PID stored in self.child is still valid, etc.
-            // If this function were fallible, we'd return an error, but
-            // since it isn't, all we can do is log an error and proceed
-            // with the wait().
-            error!(error = %e, "failed to SIGKILL; subsequent wait() might fail or wait for wrong process");
+
+        let res = tokio::time::timeout_at(timeout_at, read_page(&mut stdout, &mut page_buf))
+            .await
+            .map_err(|_elapsed| StdoutTaskError::ReadTimeout)
+            .and_then(|x| x);
+
+        match res {
+            Ok(()) => {
+                let page = Bytes::copy_from_slice(&page_buf[..8192]);
+                // we don't care about the result, because the caller could be gone
+                drop(completion.send(Ok(page)));
+                // if we did read more than one page, drain will move the one byte to the front, so we
+                // can continue with that next round.
+                page_buf.drain(..8192);
```

Comment:
I am actually quite unsure on this because I didn't read `impl bytes::BufMut for Vec<u8>` how exactly does it advance, will it override already received byte or not. But this doesn't happen in practice, so not a huge blocker for reading this.

Reply:
Almost forgot, adding a test case for this.

Reply:
Added.